swift/ql/lib/codeql/swift/elements/decl/CapturedDecl.qll

```diff
@@ -1,5 +1,23 @@
 private import codeql.swift.generated.decl.CapturedDecl
+private import codeql.swift.elements.Callable
+private import codeql.swift.elements.expr.DeclRefExpr
 
+/**
+ * A captured variable or function parameter in the scope of a closure.
+ */
 class CapturedDecl extends Generated::CapturedDecl {
   override string toString() { result = this.getDecl().toString() }
+
+  /**
+   * Gets the closure or function that captures this variable.
+   */
+  Callable getScope() { result.getACapture() = this }
+
+  /**
+   * Get an access to this capture within the scope of its closure.
+   */
+  DeclRefExpr getAnAccess() {
+    result.getEnclosingCallable() = this.getScope() and
+    result.getDecl() = this.getDecl()
```

Comment:
This looks like something that is a good candidate for [folding](https://codeql.github.com/docs/writing-codeql-queries/troubleshooting-query-performance/#fold-predicates).

Reply:
Interesting. So is any predicate of the form

```prolog
p(X, Y) :-
  f₁(X) = g₁(Y),
        ⋮
  fₖ(X) = gₖ(Y).
```

a candidate to be transformed into

```prolog
pʹ(X, Y) :-
  f(X, Z₁, …, Zₖ),
  g(Y, Z₁, …, Zₖ).

f(X, f₁(X), …, fₖ(X)).
g(Y, g₁(Y), …, gₖ(Y)).
```

?

Reply:
They're certainly candidates, yes. The question then comes down to whether or not it's worth materializing a predicate (the `locInfo` predicate in the docs) to do the more optimal join.

Reply:
Actually, in this case we get the same code before and after folding:

<details>
<summary>Before</summary>

```prolog
quick_eval#ff(/* CapturedDecl::CapturedDecl */ Synth#5f134a93::Synth::TElement this,
              /* DeclRefExpr::DeclRefExpr */ Synth#5f134a93::Synth::TElement result)
:-
  exists(/* CapturedDecl::CapturedDecl */ Synth#5f134a93::Synth::TElement call_result#2 |
    exists(/* Decl::Decl */ Synth#5f134a93::Synth::TElement call_result#3 |
      exists(/* Callable::Callable */ Synth#5f134a93::Synth::TElement call_result |
        exists(cached dontcare int _ |
          Callable#679a1d42::Generated::Callable::getImmediateCapture#1#dispred#fff(call_result,
                                                                                    _,
                                                                                    call_result#2)
        ),
        AstNode#7016758f::AstNode::getEnclosingCallable#0#dispred#bb(result,
                                                                     call_result)
      ),
      DeclRefExpr#d7ba79fe::Generated::DeclRefExpr::getDecl#0#dispred#ff(result,
                                                                         call_result#3),
      CapturedDecl#0e245f96::Generated::CapturedDecl::getDecl#0#dispred#ff(this,
                                                                           call_result#3)
    ),
    Element#e67432df::Generated::Element::resolve#ff(call_result#2, this)
  )
.
```

</details>

<details>
<summary>After</summary>

```prolog
quick_eval#ff(/* CapturedDecl::CapturedDecl */ Synth#5f134a93::Synth::TElement this,
              /* DeclRefExpr::DeclRefExpr */ Synth#5f134a93::Synth::TElement result)
:-
  exists(/* CapturedDecl::CapturedDecl */ Synth#5f134a93::Synth::TElement call_result#2 |
    exists(/* Decl::Decl */ Synth#5f134a93::Synth::TElement d |
      exists(/* Callable::Callable */ Synth#5f134a93::Synth::TElement c |
        exists(cached dontcare int _ |
          Callable#679a1d42::Generated::Callable::getImmediateCapture#1#dispred#fff(c,
                                                                                    _,
                                                                                    call_result#2)
        ),
        AstNode#7016758f::AstNode::getEnclosingCallable#0#dispred#bb(result, c)
      ),
      DeclRefExpr#d7ba79fe::Generated::DeclRefExpr::getDecl#0#dispred#ff(result,
                                                                         d),
      CapturedDecl#0e245f96::Generated::CapturedDecl::getDecl#0#dispred#ff(this,
                                                                           d)
    ),
    Element#e67432df::Generated::Element::resolve#ff(call_result#2, this)
  )
.
```

From this code:

```codeql
  …
  DeclRefExpr getAnAccess() {
    exists(Callable c, Decl d |
      getAnAccess_accessInfo(result, c, d) and
      getAnAccess_captureInfo(this, c, d)
    )
  }
}


private predicate getAnAccess_accessInfo(DeclRefExpr e, Callable c, Decl d) {
  e.getEnclosingCallable() = c and
  e.getDecl() = d
}
private predicate getAnAccess_captureInfo(CapturedDecl e, Callable c, Decl d) {
  e.getScope() = c and
  e.getDecl() = d
}

```

</details>

Reply:
No, it means that you need to add a `pragma[nomagic]` to the new predicate 😄 (and we should really update the documentation to do the same).

Reply:
Hm, makes no difference in tuple count, or rather it takes one second longer after folding, possibly due to having to do more steps and reify two predicates with roughly the same tuple count as the rest.

<details><summary>Before</summary>
<p>

```
[2023-04-03 13:39:12] Evaluated non-recursive predicate quick_eval#query#ffffffffffffff@da28b8tp in 47ms (size: 25883).
Evaluated relational algebra for predicate quick_eval#query#ffffffffffffff@da28b8tp with tuple counts:
        25883   ~6%    {14} r1 = JOIN _Locatable#2e8d6ddb::Generated::Locatable::getLocation#0#dispred#bf_Location#d34fd4c8::Location::has__#shared WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.1, "(no string representation)", Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Lhs.2, "(no string representation)", Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5
                   
        25883   ~0%    {14} r2 = JOIN _Locatable#2e8d6ddb::Generated::Locatable::getLocation#0#dispred#bf_quick_eval#ff#shared WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.1, "(no string representation)", Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5, Lhs.2, "(no string representation)", "", 0, 0, 0, 0
                   
        25883   ~0%    {14} r3 = SCAN quick_eval#ff OUTPUT In.0, "(no string representation)", "", 0, 0, 0, 0, In.1, "(no string representation)", "", 0, 0, 0, 0
            0   ~0%    {14} r4 = r3 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__Locatable#2e8d6ddb::Generated::Locat__#antijoin_rhs(Lhs.0, Lhs.7)
                   
        25883   ~0%    {14} r5 = r2 UNION r4
            0   ~0%    {14} r6 = r5 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__Locatable#2e8d6ddb::Generated::Locat__#antijoin_rhs#1(Lhs.0, Lhs.7, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6)
                   
        25883   ~6%    {14} r7 = r1 UNION r6
            0   ~0%    {14} r8 = r7 AND NOT project#CapturedDecl#2a58ca28::CapturedDecl::toString#0#dispred#ff(Lhs.0)
                   
        25883   ~3%    {12} r9 = JOIN _Locatable#2e8d6ddb::Generated::Locatable::getLocation#0#dispred#bf_Location#d34fd4c8::Location::has__#shared WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5
                   
        25883   ~6%    {12} r10 = JOIN _Locatable#2e8d6ddb::Generated::Locatable::getLocation#0#dispred#bf_quick_eval#ff#shared WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.1, Lhs.2, Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5, "", 0, 0, 0, 0
                   
        25883   ~0%    {12} r11 = SCAN quick_eval#ff OUTPUT In.0, In.1, "", 0, 0, 0, 0, "", 0, 0, 0, 0
            0   ~0%    {12} r12 = r11 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__Locatable#2e8d6ddb::Generated::Locat__#antijoin_rhs(Lhs.0, Lhs.1)
                   
        25883   ~6%    {12} r13 = r10 UNION r12
            0   ~0%    {12} r14 = r13 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__Locatable#2e8d6ddb::Generated::Locat__#antijoin_rhs#1(Lhs.0, Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6)
                   
        25883   ~3%    {12} r15 = r9 UNION r14
                   
        25883   ~0%    {14} r16 = JOIN r15 WITH CapturedDecl#2a58ca28::CapturedDecl::toString#0#dispred#ff ON FIRST 1 OUTPUT Lhs.0, Rhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.1, "(no string representation)", Lhs.7, Lhs.8, Lhs.9, Lhs.10, Lhs.11
                   
        25883   ~0%    {14} r17 = r8 UNION r16
            0   ~0%    {14} r18 = r17 AND NOT project#DeclRefExpr#5c1495fd::DeclRefExpr::toString#0#dispred#ff(Lhs.7)
                   
        25883   ~0%    {13} r19 = JOIN r15 WITH CapturedDecl#2a58ca28::CapturedDecl::toString#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Lhs.0, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Lhs.8, Lhs.9, Lhs.10, Lhs.11, Rhs.1
                   
        25883   ~6%    {13} r20 = JOIN _Locatable#2e8d6ddb::Generated::Locatable::getLocation#0#dispred#bf_Location#d34fd4c8::Location::has__#shared WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.2, Lhs.1, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5, "(no string representation)"
                   
        25883   ~4%    {13} r21 = JOIN _Locatable#2e8d6ddb::Generated::Locatable::getLocation#0#dispred#bf_quick_eval#ff#shared WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.2, Lhs.1, Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5, "", 0, 0, 0, 0, "(no string representation)"
                   
        25883   ~0%    {13} r22 = SCAN quick_eval#ff OUTPUT In.1, In.0, "", 0, 0, 0, 0, "", 0, 0, 0, 0, "(no string representation)"
            0   ~0%    {13} r23 = r22 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__Locatable#2e8d6ddb::Generated::Locat__#antijoin_rhs(Lhs.1, Lhs.0)
                   
        25883   ~4%    {13} r24 = r21 UNION r23
            0   ~0%    {13} r25 = r24 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__Locatable#2e8d6ddb::Generated::Locat__#antijoin_rhs#1(Lhs.1, Lhs.0, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6)
                   
        25883   ~6%    {13} r26 = r20 UNION r25
            0   ~0%    {13} r27 = r26 AND NOT project#CapturedDecl#2a58ca28::CapturedDecl::toString#0#dispred#ff(Lhs.1)
                   
        25883   ~0%    {13} r28 = r19 UNION r27
        25883   ~0%    {14} r29 = JOIN r28 WITH DeclRefExpr#5c1495fd::DeclRefExpr::toString#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Lhs.12, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.0, Rhs.1, Lhs.7, Lhs.8, Lhs.9, Lhs.10, Lhs.11
                   
        25883   ~0%    {14} r30 = r18 UNION r29
                       return r30
```

</details> 

<details><summary>After</summary>
<p>

```
[2023-04-03 13:40:01] Evaluated non-recursive predicate _CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs_CapturedDecl#2a58ca28::getAnAccess__#shared#1@3f5fb2hf in 13ms (size: 25883).
Evaluated relational algebra for predicate _CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs_CapturedDecl#2a58ca28::getAnAccess__#shared#1@3f5fb2hf with tuple counts:
        25883   ~1%    {7} r1 = JOIN _CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs_CapturedDecl#2a58ca28::getAnAccess__#shared WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.2, Lhs.1, Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5
                   
        23246   ~4%    {3} r2 = SCAN CapturedDecl#2a58ca28::getAnAccess_captureInfo#3#fff OUTPUT In.1, In.2, In.0
        25883   ~0%    {7} r3 = JOIN r2 WITH CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs ON FIRST 2 OUTPUT Rhs.2, Lhs.2, "", 0, 0, 0, 0
            0   ~0%    {7} r4 = r3 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__CapturedDecl#2a58ca28::getAnAccess_a__#antijoin_rhs(Lhs.1, Lhs.0)
                   
        25883   ~1%    {7} r5 = r1 UNION r4
        25883   ~6%    {8} r6 = JOIN r5 WITH Locatable#2e8d6ddb::Generated::Locatable::getLocation#0#dispred#bf ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.0, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6
                       return r6

[2023-04-03 13:40:01] Evaluated non-recursive predicate _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__CapturedDecl#2a58ca28::getAnAccess_a__#antijoin_rhs#1@1a1b3bh4 in 4ms (size: 25883).
Evaluated relational algebra for predicate _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__CapturedDecl#2a58ca28::getAnAccess_a__#antijoin_rhs#1@1a1b3bh4 with tuple counts:
        25883  ~0%    {7} r1 = JOIN _CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs_CapturedDecl#2a58ca28::getAnAccess__#shared#1 WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7
                      return r1

[2023-04-03 13:40:01] Evaluated non-recursive predicate quick_eval#query#ffffffffffffff@ecca813h in 57ms (size: 25883).
Evaluated relational algebra for predicate quick_eval#query#ffffffffffffff@ecca813h with tuple counts:
        25883   ~0%    {14} r1 = JOIN _CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs_CapturedDecl#2a58ca28::getAnAccess__#shared#1 WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.1, "(no string representation)", Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Lhs.2, "(no string representation)", Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5
                   
        25883   ~0%    {14} r2 = JOIN _CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs_CapturedDecl#2a58ca28::getAnAccess__#shared WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.1, "(no string representation)", Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5, Lhs.2, "(no string representation)", "", 0, 0, 0, 0
                   
        23246   ~4%    {3} r3 = SCAN CapturedDecl#2a58ca28::getAnAccess_captureInfo#3#fff OUTPUT In.1, In.2, In.0
                   
        25883   ~3%    {14} r4 = JOIN r3 WITH CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs ON FIRST 2 OUTPUT Lhs.2, "(no string representation)", "", 0, 0, 0, 0, Rhs.2, "(no string representation)", "", 0, 0, 0, 0
            0   ~0%    {14} r5 = r4 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__CapturedDecl#2a58ca28::getAnAccess_a__#antijoin_rhs(Lhs.0, Lhs.7)
                   
        25883   ~0%    {14} r6 = r2 UNION r5
            0   ~0%    {14} r7 = r6 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__CapturedDecl#2a58ca28::getAnAccess_a__#antijoin_rhs#1(Lhs.0, Lhs.7, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6)
                   
        25883   ~0%    {14} r8 = r1 UNION r7
            0   ~0%    {14} r9 = r8 AND NOT project#CapturedDecl#2a58ca28::CapturedDecl::toString#0#dispred#ff(Lhs.0)
                   
        25883   ~0%    {12} r10 = JOIN _CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs_CapturedDecl#2a58ca28::getAnAccess__#shared#1 WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5
                   
        25883   ~0%    {12} r11 = JOIN _CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs_CapturedDecl#2a58ca28::getAnAccess__#shared WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.1, Lhs.2, Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5, "", 0, 0, 0, 0
                   
        25883   ~3%    {12} r12 = JOIN r3 WITH CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs ON FIRST 2 OUTPUT Lhs.2, Rhs.2, "", 0, 0, 0, 0, "", 0, 0, 0, 0
            0   ~0%    {12} r13 = r12 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__CapturedDecl#2a58ca28::getAnAccess_a__#antijoin_rhs(Lhs.0, Lhs.1)
                   
        25883   ~0%    {12} r14 = r11 UNION r13
            0   ~0%    {12} r15 = r14 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__CapturedDecl#2a58ca28::getAnAccess_a__#antijoin_rhs#1(Lhs.0, Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6)
                   
        25883   ~0%    {12} r16 = r10 UNION r15
                   
        25883   ~0%    {14} r17 = JOIN r16 WITH CapturedDecl#2a58ca28::CapturedDecl::toString#0#dispred#ff ON FIRST 1 OUTPUT Lhs.0, Rhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.1, "(no string representation)", Lhs.7, Lhs.8, Lhs.9, Lhs.10, Lhs.11
                   
        25883   ~0%    {14} r18 = r9 UNION r17
            0   ~0%    {14} r19 = r18 AND NOT project#DeclRefExpr#5c1495fd::DeclRefExpr::toString#0#dispred#ff(Lhs.7)
                   
        25883   ~0%    {13} r20 = JOIN r16 WITH CapturedDecl#2a58ca28::CapturedDecl::toString#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Lhs.0, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Lhs.8, Lhs.9, Lhs.10, Lhs.11, Rhs.1
                   
        25883   ~0%    {13} r21 = JOIN _CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs_CapturedDecl#2a58ca28::getAnAccess__#shared#1 WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.2, Lhs.1, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5, "(no string representation)"
                   
        25883   ~0%    {13} r22 = JOIN _CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs_CapturedDecl#2a58ca28::getAnAccess__#shared WITH Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff ON FIRST 1 OUTPUT Lhs.2, Lhs.1, Rhs.1, Rhs.2, Rhs.3, Rhs.4, Rhs.5, "", 0, 0, 0, 0, "(no string representation)"
                   
        25883   ~4%    {13} r23 = JOIN r3 WITH CapturedDecl#2a58ca28::getAnAccess_accessInfo#3#fff_120#join_rhs ON FIRST 2 OUTPUT Rhs.2, Lhs.2, "", 0, 0, 0, 0, "", 0, 0, 0, 0, "(no string representation)"
            0   ~0%    {13} r24 = r23 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__CapturedDecl#2a58ca28::getAnAccess_a__#antijoin_rhs(Lhs.1, Lhs.0)
                   
        25883   ~0%    {13} r25 = r22 UNION r24
            0   ~0%    {13} r26 = r25 AND NOT _Location#d34fd4c8::Location::hasLocationInfo#5#dispred#ffffff__CapturedDecl#2a58ca28::getAnAccess_a__#antijoin_rhs#1(Lhs.1, Lhs.0, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6)
                   
        25883   ~0%    {13} r27 = r21 UNION r26
            0   ~0%    {13} r28 = r27 AND NOT project#CapturedDecl#2a58ca28::CapturedDecl::toString#0#dispred#ff(Lhs.1)
                   
        25883   ~0%    {13} r29 = r20 UNION r28
        25883   ~0%    {14} r30 = JOIN r29 WITH DeclRefExpr#5c1495fd::DeclRefExpr::toString#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Lhs.12, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.0, Rhs.1, Lhs.7, Lhs.8, Lhs.9, Lhs.10, Lhs.11
                   
        25883   ~0%    {14} r31 = r19 UNION r30
                       return r31
```

</p>
</details>